tpdatasrc/tpgamefiles/scr/spell_utils.py

```diff
@@ -0,0 +1,275 @@
+    elif duration < 100: # 10 minutes
+        return "{} rounds".format(duration)
+    elif duration < 1200: # 2 hours
+        return "{} minutes".format(duration/10)
+    elif duration < 28800: # 2 days
+        return "{} hours".format(duration/600)
+    elif duration < 864000: # 2 months
+        return "{} days".format(duration/14400)
+    else:
+        return "{} months".format(duration/432000)
+
+### Standard Hooks invoked with AddHook ###
+
+#[pytonModifier].AddHook(ET_OnGetTooltip, EK_NONE, spell_utils.spellTooltip, ())
+def spellTooltip(attachee, args, evt_obj):
+    name = spellName(args.get_arg(0))
+    duration = spellTime(args.get_arg(1))
+    evt_obj.append("{} ({})".format(name, duration))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnGetEffectTooltip, EK_NONE, spell_utils.spellEffectTooltip, ())
+def spellEffectTooltip(attachee, args, evt_obj):
+    key = spellKey(args.get_arg(0))
+    duration = spellTime(args.get_arg(1))
+    evt_obj.append(key, -2, " ({})".format(duration))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Query, EK_Q_Critter_Has_Spell_Active, spell_utils.queryActiveSpell, ())
+def queryActiveSpell(attachee, args, evt_obj):
+    spellPacket = tpdp.SpellPacket(args.get_arg(0))
+    if evt_obj.data1 == spellPacket.spell_enum:
+        evt_obj.return_val = 1
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Killed, spell_utils.spellKilled, ())
+def spellKilled(attachee, args, evt_obj):
+    args.remove_spell()
+    args.remove_spell_mod()
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Spell_End, spell_utils.spellEnd, ())
+def spellEnd(attachee, args, evt_obj):
+    print "{} SpellEnd".format(spellName(args.get_arg(0)))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Concentration_Broken, spell_utils.checkRemoveSpell, ())
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Dismiss_Spells, spell_utils.checkRemoveSpell, ())
+def checkRemoveSpell(attachee, args, evt_obj):
+    if evt_obj.data1 == args.get_arg(0):
+        args.remove_spell()
+        args.remove_spell_mod()
+    return 0
+
+#[pytonModifier].AddHook(ET_OnConditionAdd, EK_NONE, spell_utils.addConcentration, ())
+def addConcentration(attachee, args, evt_obj):
+    spellPacket = tpdp.SpellPacket(args.get_arg(0))
+    spellPacket.caster.condition_add_with_args('sp-Concentrating', args.get_arg(0))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnConditionAdd, EK_NONE, spell_utils.addDimiss, ())
+def addDimiss(attachee, args, evt_obj):
+    spellPacket = tpdp.SpellPacket(args.get_arg(0))
+    spellPacket.caster.condition_add_with_args('Dismiss', args.get_arg(0))
+    return 0
+
+#[pytonModifier].AddHook(ET_OnD20Signal, EK_S_Temporary_Hit_Points_Removed, spell_utils.removeTempHp, ())
+# needed in combination with condition_add_with_args('Temporary_Hit_Points', 0, spell.duration, tempHpAmount)
+def removeTempHp(attachee, args, evt_obj):
+    attachee.d20_send_signal(S_Spell_End, 'Temporary_Hit_Points')
```

Comment:
This is unspecific to any particular spell... could cause issues if multiple spells add temp HP this way.
@Sagenlicht

Reply:
This should only be triggered if Temp HP value <= 0, no matter from what sources Temp HP were originally granted. My bad, if I did not make this clear.

I just noticed, when I use attachee.condition_add_with_args('Temporary_Hit_Points', 0,  duration, tempHP) the condition does not automatically gets removed when temp HP reach 0 but requires an spell_end signal.

When TempHP falls to or below 0, a signal is triggered (S_Temporary_Hit_Points_Removed) but it actually requires the spell triggering then the removal of the Temporary_Hit_Points condition by doing the added line in spell_utils.

It basically looks like:
def righteousFurySpellOnConditionAdd(attachee, args, evt_obj):
    attachee.condition_add_with_args('Temporary_Hit_Points', 0, 600, args.get_arg(2)) #temp HP value is passed by spell (arg3), duration is up to 1 hour;
    return 0
def righteousFurySpellRemoveTempHp(attachee, args, evt_obj):
    attachee.d20_send_signal(S_Spell_End, 'Temporary_Hit_Points')
    return 0
righteousFurySpell.AddHook(ET_OnConditionAdd, EK_NONE, righteousFurySpellOnConditionAdd,())
righteousFurySpell.AddHook(ET_OnD20Signal, EK_S_Temporary_Hit_Points_Removed, righteousFurySpellRemoveTempHp, ())

I now want to replace the signal handling by:
righteousFurySpell.AddHook(ET_OnD20Signal, EK_S_Temporary_Hit_Points_Removed, spell_utils.removeTempHp, ())

Reply:
Are you sure it's only triggered by that?
Also note that attache.d20_send_signal(S_Spell_End, 'blah blah') is meaningless, and is parsed internally to attache.d20_send_signal(S_Spell_End, 0, 0).
I believe that if you cast two different spells with this temp HP scripting, removing the temp HP of one (due to damage) will end the other.

Reply:
No, I am not 100% sure, but I tested a lot with Temporary Hit Points and tried to understand what the game actually does. I use the vanilla condition 'Temporary_Hit_Points' (https://github.com/GrognardsFromHell/DllDocumentation/wiki/Conditions-(List)-page-2-M-Z#temporary_hit_points) for all my spells that grant Temp HP. Basically I add the condition as a secondary condition. I belive this exactly what the vanilla spell Aid does.

When the Temp HP are reduced to 0 or below the 'Temporary_Hit_Points' condition does not get removed (which I originally hoped it would) and I digged around in the C code to understand what the Aid spell does. 

Both the spell https://github.com/GrognardsFromHell/DllDocumentation/wiki/Conditions-(List)-Spells#sp-aid and 
https://github.com/GrognardsFromHell/DllDocumentation/wiki/Conditions-(List)-Spells-page-2#sp-false-life 
react to the DK_SIG_Temporary_Hit_Points_Removed signal.

I did not test what happens if more than one source grants temp HP, I did not test if the condition can be applied more than once, but I tested and know 100% that the signal will only remove the Temporary_Hit_Points condition and not my original spell condition. For example Righteous Fury grants + 4 Str and Temp HP, when the Teemp HP is gone the condition gets removed but the original spell condition stays and still grants +4 Str.

Reply:
As you said, both Aid and False life react to that signal... by sending d20_send_signal(S_Spell_End, **spell.id**, 0). If they hadn't specified the spell ID, they'd also end each other's temp HP.

Reply:
Ok I think, I understood what you want me to do.
You basically want me to add the spell.id to the condition, Will do that and modifiy the spell_utils as well then.
Instead of:
`attachee.condition_add_with_args('Temporary_Hit_Points', 0, duration, tempHpAmount)`
it should be:
`attachee.condition_add_with_args('Temporary_Hit_Points', spell_id, duration, tempHpAmount)`

and then in the spell_utils:
`def removeTempHp(attachee, args, evt_obj):`
` attachee.d20_send_signal(S_Spell_End, 'Temporary_Hit_Points', args.get_arg(0))`
`    return 0`

Will do that.

Reply:
After changing it to the above, the Condition does not get removed anymore when the signal is send :( 

Will need to spend some more time on this then.

Reply:
`attachee.d20_send_signal(S_Spell_End,'Temporary_Hit_Points', args.get_arg(0))`
This is the problem part. Again, the string arg is meaningless and will be parsed by the engine to 0. Remove it and it should work.

Reply:
Ok, wow guess brain afk, got it. Will fix it and ty.

Reply:
I should probably make it throw an error in such cases :)